gourmet/timer.py

```diff
@@ -1,99 +1,128 @@
-            s.set_width_chars(2)
-            s.set_value(1)
-            s.set_value(0)
+            spinner.connect('changed', self.val_changed_cb)
+            spinner.val_change_is_changing_entry = False
+            spinner.set_width_chars(2)
+
+    def set_time(self, val: int) -> None:
+        """Update the spinners on tick.
+
+        Ran on every tick, the value is split into hours, minutes, seconds, and
+        each spinner is set, to show the time going down.
+        """
+        val = max(0, val)
+        self.hoursSpin.set_value(val // 3600)
+        val = val % 3600
+        self.minutesSpin.set_value(val // 60)
+        self.secondsSpin.set_value(val % 60)
+
+    def get_time(self) -> int:
+        """Get the time to run the timer for, in seconds"""
+        return self.hoursSpin.get_value()*3600 + self.minutesSpin.get_value()*60 + self.secondsSpin.get_value()
 
-    def set_time (self, s):
-        s = int(s)
-        if s < 0: s = 0
-        #self.hoursSpin.set_text(self.pad_n(s / 3600))
-        self.hoursSpin.set_value(s / 3600)
-        s = s % 3600
-        #self.minutesSpin.set_text(self.pad_n(s / 60))
-        self.minutesSpin.set_value(s / 60)
-        #self.secondsSpin.set_text(self.pad_n(s % 60))
-        self.secondsSpin.set_value(s % 60)
-
-    def val_changed_cb (self, widg):
+    def val_changed_cb(self, widg: Gtk.SpinButton) -> None:
+        """On input callback to set the values to be always two digits"""
         if not widg.val_change_is_changing_entry:
             widg.val_change_is_changing_entry = True
-            widg.set_text(self.pad_n(int(widg.get_value())))
+            widg.set_text(widg.get_text().zfill(2))
             widg.val_change_is_changing_entry = False
 
-    def pad_n (self, int):
-        s = str(int)
-        if len(s)==1: return '0'+s
-        else: return s
+    def tick(self) -> bool:
+        """Run the timer
 
-    def get_time (self):
-        return self.hoursSpin.get_value()*3600 + self.minutesSpin.get_value()*60 + self.secondsSpin.get_value()
+        This is done in two steps: first we compute the time elapsed between two
+        ticks of the loop, and then we subtract it from the timer time, which is
+        then updated.
+
+        Returns a bool to notify the event-loop on whether the timer is done.
+        """
+        if self.is_running:
+            if self.previous_iter_time is None:
+                self.previous_iter_time = time.time()
+
+            now = time.time()
+            elapsed = now - self.previous_iter_time
+            self.previous_iter_time = now
 
-    # Methods to run the timer...
-    def tick (self):
-        if self.running:
-            elapsed = time.time() - self.running
-            t = self.start_time - elapsed
-            #t = self.get_time() - 1
-            self.set_time(t)
-            if t<=0:
+            t = self.get_time()
+            remaining = t - elapsed
+            self.set_time(remaining)
```

Comment:
time has subsecond precision and `time.time()` returns a float but `set_time` is supposed to take an int. Maybe `set_time` takes a float as well and we round or truncate it?

Reply:
but we want to keep the interal value (`previous_iter_time`) at subsecond precision. Best not to round there.

Reply:
Yes, it's a float, my bad